### PR TITLE
[#3378] Fix formstep url in logic rules of copied form

### LINKED ISSUE #3378: Form copying with logic rules that contain: "Inschakelen vanaf stap:" breaks Logic tab in new copy
### Product versie / Product version

2.2.x - 2.3.0
Haarlemmermeer: 28

### Omschrijf het probleem / Describe the bug

If a form containing logic rules that are enabled from a certain step, the copy of the form will have a broken Logic tab as it seems the "Inschakelen vanaf stap" still pointing to the original form:

```json
 {
        "uuid": "de3cc210-aa72-437c-9e37-8aeec24ebe5b",
        "url": "http://localhost/api/v2/forms/de3cc210-aa72-437c-9e37-8aeec24ebe5b/logic-rules",
        "form": "http://localhost/api/v2/forms/1dacaf4f-bc68-4197-8c7b-84ad7ec913c1",
        "json_logic_trigger": true,
        "description": "",
        "order": 3,
        "trigger_from_step": "http://localhost/api/v2/forms/1dacaf4f-bc68-4197-8c7b-84ad7ec913c1/steps/72d4bdb5-90f0-4d89-9c3b-1bf3373c49d7",
        "actions": [
            {
                "component": "",
                "variable": "lijstvoorkvkvestiginen",
                "form_step": "",
                "form_step_uuid": null,
                "action": {
                    "type": "variable",
                    "value": ...
                }
            }
        ],
        "is_advanced": true
    }
```

json in copied form, notice the: `"trigger_from_step": "http://localhost/api/v2/forms/1dacaf4f-bc68-4197-8c7b-84ad7ec913c1/steps/72d4bdb5-90f0-4d89-9c3b-1bf3373c49d7"` instead of `http://localhost/api/v2/forms/1968d36e-8932-43ff-8a83-8c22a2ad97dd`
```json
    {
        "uuid": "5fa0ea6f-c0d8-43b5-adb0-ad2584208bda",
        "url": "http://localhost/api/v2/forms/5fa0ea6f-c0d8-43b5-adb0-ad2584208bda/logic-rules",
        "form": "http://localhost/api/v2/forms/1968d36e-8932-43ff-8a83-8c22a2ad97dd",
        "json_logic_trigger": true,
        "description": "",
        "order": 3,
        "trigger_from_step": "http://localhost/api/v2/forms/1dacaf4f-bc68-4197-8c7b-84ad7ec913c1/steps/72d4bdb5-90f0-4d89-9c3b-1bf3373c49d7",
        "actions": [
            {
                "component": "",
                "variable": "lijstvoorkvkvestiginen",
                "form_step": "",
                "form_step_uuid": null,
                "action": {
                    "type": "variable",
                    "value": ...
                }
            }
        ],
        "is_advanced": true
    }
```

Importing the form does work.

### PR DESCRIPTION
**Pull request by vaszig**
Fixes #3378

### CODE REVIEW

File: src/openforms/forms/tests/admin/test_form.py
Comment (SilviaAmAm) [inline src/openforms/forms/tests/admin/test_form.py:1]: I think it would be better to restructure the tests as follows:
- test_form_admin_copy (the same as the old one): this can be the same as what was there before => it checks the redirect after the POST action, it checks the name/uuids/logic rules/auth plugins/...
- `test_copy_form_with_trigger_from_step_in_logic` (or something like that...) that could have the decorator `@tag("gh-3378")` that creates a form with a logic rule with a trigger step and then only checks that the trigger step has been copied properly.
- Remove the new `setUp`

In this way the tests are more self contained, and the test for this issue is more specific to the bug (and it also doesn't duplicate checks that are already done by the other test).

File: src/openforms/forms/models/form.py [hunk 1]
```diff
@@ -475,6 +475,12 @@ def copy(self):
             logic.pk = None
             logic.uuid = _uuid.uuid4()
             logic.form = copy
+
+            if logic.trigger_from_step:
+                logic.trigger_from_step = logic.form.formstep_set.get(
+                    order=logic.trigger_from_step.order
+                )
```
[the remark this answers is not in the document]
Comment (vaszig) [reply src/openforms/forms/models/form.py:482]: We can't use `refresh_from_db()` directly as formsteps is a queryset (`MultilingualOrderedModelQuerySet`), right?
So we will need a for loop again, or maybe you mean something else that I didn't get?

File: src/openforms/forms/tests/admin/test_form.py
Comment (vaszig) [reply src/openforms/forms/tests/admin/test_form.py:1]: Thanks for the suggestion, it's indeed a neater solution!I will update the test file.

File: src/openforms/forms/models/form.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openforms/forms/models/form.py:482]: Ahh yes you are right, I missed that it is a query set

[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/openforms/forms/models/form.py:482]: A small optimization would be:

(before the `for logic in ...` loop)

```py
steps_by_order = {
    step.order: step for step in logic.form.formstep_set.all()
}
```

and then in the loop:

```py
logic.trigger_from_step = steps_by_order[logic.trigger_from_step.order]
```
so you replace the amount of queries from `n`  to `1`.

This is probably a bit of premature optimization so consider this just information, no need to update the PR.

File: src/openforms/forms/tests/admin/test_form.py [hunk 2]
```diff
@@ -550,6 +550,26 @@ def test_form_admin_copy(self):
...
+        form_step = FormStepFactory.create(form=form, form_definition__is_reusable=True)
+        FormLogicFactory.create(form=form, trigger_from_step=form_step)
+
+        # React UI renders this input, so simulate it in a raw POST call
+        self.client.post(admin_url, data={"_copy": "Copy"})
+
+        copied_form = Form.objects.get(slug=_("{slug}-copy").format(slug=form.slug))
+        copied_logic = copied_form.formlogic_set.get()
+
+        self.assertEqual(copied_logic.trigger_from_step, copied_form.formstep_set.get())
```
Comment (sergei-maertens) [inline src/openforms/forms/tests/admin/test_form.py:571]: I think we can simplify this test case by only keeping what's relevant for this particular issue:

```suggestion
    @tag("gh-3378")
    def test_copy_form_with_trigger_from_step_in_logic(self):
        user = SuperUserFactory.create()
        self.client.force_login(user)

        form_step = FormStepFactory.create()
        FormLogicFactory.create(form=form_step.form, trigger_from_step=form_step)
        admin_url = reverse("admin:forms_form_change", args=(form_step.form.pk,))

        # React UI renders this input, so simulate it in a raw POST call
        self.client.post(admin_url, data={"_copy": "Copy"})

        copied_form = Form.objects.exclude(pk=form_step.form.pk).get()
        copied_step = FormStep.objects.exclude(pk=form_step.pk).get()
        copied_logic = copied_form.formlogic_set.get()

        self.assertEqual(copied_logic.trigger_from_step, copied_step)
```

Authentication backends or any form-specific configuration is completely irrelevant here - and we can explicitly query for the copied objects (form, step, logic) since the test data is set up so that there's only 1 form to begin with and after the copy there should only be one additional form. Using the `queryset.get()` approach enforces this and prevents surprises/mistakes in test data setup.

**Review by sergei-maertens — APPROVED** [review]

**Comment by sergei-maertens** [comment]
Backports:

* `stable/2.3.x`: 8512fe4a566cc4909ed8f1072a0410785d49b6df
* `stable/2.2.x`: 361dc2d8369a41bbc3ccd641c79c771527253664
* `stable/2.1.x`: d128cb1a368675b6b9a3057dd6cc142b6f2860f9
* `stable/2.0.x`: 9d5b35f508d1baaf7137084e32a5fefb270b08bf
